allennlp/training/trainer.py

```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
-                    # Is the worst validation performance in past self._patience
-                    # epochs is better than current value?
-                    if self._validation_metric_decreases:
-                        should_stop = max(validation_metric_per_epoch[-self._patience:]) < this_epoch_val_metric
-                    else:
-                        should_stop = min(validation_metric_per_epoch[-self._patience:]) > this_epoch_val_metric
-                    if should_stop:
-                        logger.info("Ran out of patience.  Stopping training.")
-                        break
-                validation_metric_per_epoch.append(this_epoch_val_metric)
-
-                if self._validation_metric_decreases:
-                    is_best_so_far = this_epoch_val_metric == min(validation_metric_per_epoch)
-                else:
-                    is_best_so_far = this_epoch_val_metric == max(validation_metric_per_epoch)
-                if self._serialization_dir:
-                    self._save_checkpoint(epoch, is_best=is_best_so_far)
-
-                if self._learning_rate_scheduler:
-                    # Grim hack to determine whether the validation metric we are recording
-                    # needs to be passed to the scheduler. This is required because the
-                    # step() function of the different schedulers are (understandably)
-                    # different to ReduceLROnPlateau.
-                    if isinstance(self._learning_rate_scheduler,
-                                  torch.optim.lr_scheduler.ReduceLROnPlateau):
-                        self._learning_rate_scheduler.step(this_epoch_val_metric, epoch)
-                    self._learning_rate_scheduler.step(epoch)
-            else:
-                message_template = "Training %s : %3f "
-                for name, value in metrics.items():
-                    logger.info(message_template, name, value)
-                    if self._serialization_dir:
-                        train_log.add_scalar(name, value, epoch)
-                if self._serialization_dir:
-                    self._save_checkpoint(epoch)
-                if self._learning_rate_scheduler:
-                    if isinstance(self._learning_rate_scheduler,
-                                  torch.optim.lr_scheduler.ReduceLROnPlateau):
-                        raise ConfigurationError("The reduce_on_plateau learning rate scheduler requires "
-                                                 "a validation metric to compute the schedule and therefore "
-                                                 "must be used with a validation dataset.")
-                    self._learning_rate_scheduler.step(epoch)
+            # Train one epoch and compute metrics.
+            metrics = self._train_epoch(epoch)
+
+            # Validate and move on to the next epoch
+            if self._should_stop(epoch, metrics, validation_metric_per_epoch):
```

Comment:
I feel like this one went a bit too far - I'd probably put an explicit check for validation data here, and a call to `validate` if so, otherwise just save a checkpoint, then put the learning rate scheduler update stuff after that.  Having the `next_epoch{_no}_validation` seems a little bit much to me.